### ISSUE TITLE
append to Array of Array Fields

### ISSUE THREAD

**Issue by szymexpawlowski** [react-hook-form/react-hook-form#1548]
Hi,

I'm trying to extend functionality of [Array of Field Array](https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-x7btr) example (btw. great stuff). My use case is that I would like to append to `FieldArray` an element that has `nestedArray` so some defaults in `NestedArray` are filled. Something like below in `fieldArray.js`.

```
onClick={() => {
  append({
    name: "append",
    nestedArray: [{ field1: "field1", field2: "field2" }]
  });
}}
```  
After performing the append, `fields` variable in `nestedFieldArray.js` is an empty array so nested fileds doesn't render.

Thanks in advance.

**Comment by bluebill1049** [comment]
`append`  API doesn't support that, you will have to trigger the append inside the nested field.

**Comment by bluebill1049** [comment]
here r some more context: https://github.com/react-hook-form/react-hook-form/issues/1561

**Comment by szymexpawlowski** [comment]
@bluebill1049 it's the first time I've experienced such care from lib maintainer. awesome job!

**Comment by bluebill1049** [comment]
my pleasure @szymexpawlowski it's a meaningful thing to do for me as well.

**Comment by levrik** [comment]
@bluebill1049 Do you think we could get an easier approach to this in V7 than what is suggested here and in #1561?

**Comment by bluebill1049** [comment]
> @bluebill1049 Do you think we could get an easier approach to this in V7 than what is suggested here and in #1561?

`append` will still have its own scope (useFieldArry), but you can use `setValue('keyname', [{}....])`

**Comment by levrik** [comment]
@bluebill1049 
It would be kinda cool if `useFieldArray` could use the values from `append` as default in case of nested `useFieldArray`'s.
This was also kinda what I was expecting from it.
The problem I'm facing here is that I don't know which index the newly appended item will have so kinda difficult to get the data to that newly created `useFieldArray` usage, to insert the data? Or to somehow get access to `useFieldArray` in general.

**Comment by bluebill1049** [comment]
> @bluebill1049
> It would be kinda cool if `useFieldArray` could use the values from `append` as default in case of nested `useFieldArray`'s.
> This was also kinda what I was expecting from it.
> The problem I'm facing here is that I don't know which index the newly appended item will have so kinda difficult to get the data to that newly created `useFieldArray` usage, to insert the data? Or to somehow get access to `useFieldArray` in general.

not sure if i am following you, you want to send a CSB?

**Comment by levrik** [comment]
@bluebill1049 
I'm having a structure like this:

```jsonc
{
	"array": [
		{ "type": "flat" },
		{
			"type": "nested",
			"array": [
				{ "type": "flat" },
				// ...
			]
		}
	]
}
```

Then I have a button which can add stuff to the outermost array.
But for example also with a structure like shown in `"type": "nested"`, with another field array in it.
When pushing this type nested example to `append` the nested `"array"` will be empty.
That's noted due to the design. But I really don't know how to get either access to `append` of that nested array or pushing data to the component so it can call `append` internally after mount because after calling `append` on the outermost `"array"` I don't know about the newly created index or so.
It would be great if calling
```js
append({
			"type": "nested",
			"array": [
				{ "type": "flat" },
				// ...
			]
		})
```
would result in this nested `"array"` being supplied to `useFieldArray` as default values.

I can't just call `useFieldArray` 2 times upper in the tree. I don't know how many `"type": "nested"` I will have. This is completely dynamic.

**Comment by bluebill1049** [comment]
@ivkirill certainly not going to stop you from rolling your own hook, I think it's worth of try anyhow (if you know exactly your requirement). 

``` 
append({
  "type": "nested",
  "array": [
    { "type": "flat" },
   // ...
  ]
}); 
// doesn't work with RHF
```

```
setValue('test', [...getValues().test, { newField: 'what'  }]); // works for RhF
```

**Comment by levrik** [comment]
@bluebill1049 Mhh. I'm thinking about storing the forms state in a flat key value structure in the form and having the ordering and nesting in some global state with just the keys. This probably also should be more performant than `useFieldArray`.

**Comment by bluebill1049** [comment]
> @bluebill1049 Mhh. I'm thinking about storing the forms state in a flat key value structure in the form and having the ordering and nesting in some global state with just the keys. This probably also should be more performant than `useFieldArray` but would allow me to keep advantage of a lot of great features from `react-hook-form`.

can you take a look at this example? https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-x7btr

click on `append nested` isn't that what you after?

**Comment by levrik** [comment]
@bluebill1049 Ah, yeah. Kinda. But this requires me to pass down `getValues` down the whole tree and since the identity of `getValues` isn't stable, this caused a lot of performance issues for me. I've mentioned this on another issue about `useWatch` and `useFieldArray` yesterday or so.

**Comment by bluebill1049** [comment]
i just quickly tried v7...

```
append({
  name: 'data',
  nested: [{
    test: 'xx'
  }] 
})
```

it actual works... https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-d3k6n?file=/src/fieldArray.js

**Comment by levrik** [comment]
@bluebill1049 v7 seems to be really great 😅 fixes a lot of things around `useFieldArray`.

**Comment by bluebill1049** [comment]
@levirk wow, it works for prepending as well... with v7

**Comment by levrik** [comment]
@bluebill1049 but in general I'm having other perf issues with `useFieldArray`.
when having a list of 40 items in a nested `useFieldArray` structure:
if I remove the first item of the outermost `useFieldArray` the whole tree re-renders due to shifting array indexes. every single component gets a re-render and `React.memo` doesn't help there.
the more you have in that list and the upper you remove an item, the slower it gets. this is kinda bad for UX.
that's why I was thinking of using a flat key-value structure in the form and having the ordering and nesting stored externally.
this way I can bail out of unnecessary re-renders quite easily because the props are stable.
of course this means that data I'm getting from `submit` needs a conversion before being in the correct structure but that's fine for me.

**Comment by bluebill1049** [comment]
> @bluebill1049 but in general I'm having other perf issues with `useFieldArray`.
> when having a list of 40 items in a nested `useFieldArray` structure:
> if I remove the first item of the outermost `useFieldArray` the whole tree re-renders due to shifting array indexes. every single component gets a re-render and `React.memo` doesn't help there because the index has to be passed to the component to form the right `name` for `register`.
> the more you have in that list and the upper you remove an item, the slower it gets. this is kinda bad for UX.
> that's why I was thinking of using a flat key-value structure in the form and having the ordering and nesting stored externally.
> this way I can bail out of unnecessary re-renders quite easily because the props are stable.
> of course this means that data I'm getting from `submit` needs a conversion before being in the correct structure but that's fine for me.
> 
> so far the idea. not sure if it really solves the perf issues but I think so.

if your data size is huge, I would recommend building a targeted solution for that. I don't think hook form can prevent re-render during each action.

**Comment by levrik** [comment]
@bluebill1049 wanted to try your suggested solution with `setValue` but `getValues` gives me `undefined` as soon as I'm passing in a name. Thought this is due to the usage of `defaultValues`. Switched to using `setValue` instead on first render. Still `undefined`. If I call `getValues` without a name, the correct data is returned.

**Comment by bluebill1049** [comment]
> _Edit:_ Looks like `getValues` returns `undefined` if I'm targeting `useFieldArray` fields directly.
> If I'm trying to get the value of something inside `useFieldArray` it works fine.

you should try `getValues().test`

**Comment by levrik** [comment]
@bluebill1049 this would kinda work but I'm passing down the path as a string because I have `useFieldArray` inside `useFieldArray` inside `useFieldArray` etc etc
I mean, not impossible to get the properties accessed with the help of some library or some helper method from the string path but I'm guessing I'll really give on `react-hook-form` for this use-case and build a completely custom solution.
Thanks a lot for your patience and time. The library is great and will be continued to be used throughout our app in other places. v7 will resolve most of the issues I'm currently facing but not the perf issues coming with arrays but I think that's in the nature of arrays and there isn't a better way for `react-hook-form` to handle this.

**Comment by bluebill1049** [comment]
> @bluebill1049 this would kinda work but I'm passing down the path as a string because I have `useFieldArray` inside `useFieldArray` inside `useFieldArray` etc etc
> I mean, not impossible to get the properties accessed with the help of some library or some helper method from the string path but I'm guessing I'll really give on `react-hook-form` for this use-case and build a completely custom solution.
> Thanks a lot for your patience and time. The library is great and will be continued to be used throughout our app in other places. v7 will resolve most of the issues I'm currently facing but not the perf issues coming with arrays but I think that's in the nature of arrays and there isn't a better way for `react-hook-form` to handle this.

I will ping you when the beta version is ready for RHF 7, I think you should use V7, especially you have to deal with a lot of Field Array. In fact, subscribe to this: https://github.com/react-hook-form/react-hook-form/discussions/3714 you will get notified when the beta is out. (expected to have it ready next week)

**Comment by levrik** [comment]
@bluebill1049 Sadly I kinda have to deliver right today. I mean, not gonna make this if I'm rewriting this whole thing without `react-hook-form` now but I can't wait on something anymore. I would love to but, you know, business.
